### PR TITLE
🐛 Preserve traceback when an exception is raised in sync dependency with `yield`

### LINKED ISSUE #13022: Traceback stack does not show exact place of error
### Discussed in https://github.com/fastapi/fastapi/discussions/8428

<div type='discussions-op-text'>

<sup>Originally posted by **NewSouthMjos** December  5, 2022</sup>
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI, Depends


app = FastAPI()


def get_something_sync():
    yield True


async def get_something_async():
    yield True


@app.get('/1')
def router_func(dependency=Depends(get_something_sync)):
    raise ValueError
    return


@app.get('/2')
def router_func(dependency=Depends(get_something_async)):
    raise ValueError
    return


if __name__ == "__main__":
    uvicorn.run("main:app", host="0.0.0.0", port=5600, workers=1)
```

### Description

If dependency function uses yield, injecting dependency with sync def function (get_something_sync), the error traceback will broke - it doesn't locate the place where it was rise (should be "/app/main.py", line 24)
calling endpoint /1:
```
fastapi_dependency  | INFO:     10.77.78.83:60070 - "GET /1 HTTP/1.1" 500 Internal Server Error
fastapi_dependency  | ERROR:    Exception in ASGI application
fastapi_dependency  | Traceback (most recent call last):
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
fastapi_dependency  |     result = await app(  # type: ignore[func-returns-value]
fastapi_dependency  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
fastapi_dependency  |     return await self.app(scope, receive, send)
fastapi_dependency  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 270, in __call__
fastapi_dependency  |     await super().__call__(scope, receive, send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 124, in __call__
fastapi_dependency  |     await self.middleware_stack(scope, receive, send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 184, in __call__
fastapi_dependency  |     raise exc
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 162, in __call__
fastapi_dependency  |     await self.app(scope, receive, _send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
fastapi_dependency  |     raise exc
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
fastapi_dependency  |     await self.app(scope, receive, sender)
fastapi_dependency  |   File "/usr/local/lib/python3.11/contextlib.py", line 222, in __aexit__
fastapi_dependency  |     await self.gen.athrow(typ, value, traceback)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/fastapi/concurrency.py", line 36, in contextmanager_in_threadpool
fastapi_dependency  |     raise e
fastapi_dependency  | ValueError
```

When injecting dependency with async def function (get_something_async), the error traceback seems to be right
endpoint /2:
```
fastapi_dependency  | INFO:     10.77.78.83:60067 - "GET /2 HTTP/1.1" 500 Internal Server Error
fastapi_dependency  | ERROR:    Exception in ASGI application
fastapi_dependency  | Traceback (most recent call last):
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
fastapi_dependency  |     result = await app(  # type: ignore[func-returns-value]
fastapi_dependency  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
fastapi_dependency  |     return await self.app(scope, receive, send)
fastapi_dependency  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 270, in __call__
fastapi_dependency  |     await super().__call__(scope, receive, send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 124, in __call__
fastapi_dependency  |     await self.middleware_stack(scope, receive, send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 184, in __call__
fastapi_dependency  |     raise exc
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 162, in __call__
fastapi_dependency  |     await self.app(scope, receive, _send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
fastapi_dependency  |     raise exc
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
fastapi_dependency  |     await self.app(scope, receive, sender)
fastapi_dependency  |   File "/usr/local/lib/python3.11/contextlib.py", line 222, in __aexit__
fastapi_dependency  |     await self.gen.athrow(typ, value, traceback)
fastapi_dependency  |   File "/app/main.py", line 13, in get_something_async
fastapi_dependency  |     yield True
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
fastapi_dependency  |     raise e
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
fastapi_dependency  |     await self.app(scope, receive, send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 706, in __call__
fastapi_dependency  |     await route.handle(scope, receive, send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 276, in handle
fastapi_dependency  |     await self.app(scope, receive, send)
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 66, in app
fastapi_dependency  |     response = await func(request)
fastapi_dependency  |                ^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 235, in app
fastapi_dependency  |     raw_response = await run_endpoint_function(
fastapi_dependency  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 163, in run_endpoint_function
fastapi_dependency  |     return await run_in_threadpool(dependant.call, **values)
fastapi_dependency  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
fastapi_dependency  |     return await anyio.to_thread.run_sync(func, *args)
fastapi_dependency  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/anyio/to_thread.py", line 31, in run_sync
fastapi_dependency  |     return await get_asynclib().run_sync_in_worker_thread(
fastapi_dependency  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
fastapi_dependency  |     return await future
fastapi_dependency  |            ^^^^^^^^^^^^
fastapi_dependency  |   File "/usr/local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 867, in run
fastapi_dependency  |     result = context.run(func, *args)
fastapi_dependency  |              ^^^^^^^^^^^^^^^^^^^^^^^^
fastapi_dependency  |   File "/app/main.py", line 24, in router_func
fastapi_dependency  |     raise ValueError
fastapi_dependency  | ValueError
```

On python 3.10 there were no such a problem. 
Created repo for reproduce the problem: [https://github.com/NewSouthMjos/fastapi_dependency_test](https://github.com/NewSouthMjos/fastapi_dependency_test)
So, should the sync way to inject dependecy show exact plase of where is was raised?

### Operating System

Linux

### Operating System Details

Running in a docker

### FastAPI Version

0.88.0

### Python Version

3.11

### Additional Context

uvicorn[standard]==0.20.0</div>

### PR DESCRIPTION
**Pull request by sombek**
There is a missing of traceback when an exception raised in contextmanager_in_threadpool
Happens on Python 3.11 

I noticed the exception.traceback gets flushed after exiting the context manager.

There is an issue is opened #5740 

Edit by @Kludex : 
- Closes https://github.com/fastapi/fastapi/issues/13022

### CODE REVIEW

**Comment by MohammedAlhajji** [comment]
great workaround

**Comment by MohammedAlhajji** [comment]
great workaround. thanks!

**Comment by MrSalman333** [comment]
Nice work!

We didn't upgrade to 3.11 because of this.
Thankss

**Review by yezz123 — APPROVED** [review]

**Review by iudeen — APPROVED** [review]
Looks neat! And thanks for the tests as well!

**Comment by sombek** [comment]
@iudeen @yezz123 Usually, how long it takes you guys to merge a PR?

**Comment by yezz123** [comment]
> @iudeen @yezz123 Usually, how long it takes you guys to merge a PR?

Mainly @tiangolo will review it after adding a flag of `awaiting review` so just wait it's will be reviewed asap

**Comment by sombek** [comment]
@yezz123 Who adds the `awaiting review` flag?

**Comment by suzaku** [comment]
I patched your change with the following code and now the traceback's back. Thanks.

```python
from contextlib import asynccontextmanager as asynccontextmanager
from typing import AsyncGenerator, ContextManager, TypeVar

import anyio
from anyio import CapacityLimiter
from starlette.concurrency import run_in_threadpool as run_in_threadpool  # noqa

_T = TypeVar("_T")

# 
@asynccontextmanager
async def contextmanager_in_threadpool(
    cm: ContextManager[_T],
) -> AsyncGenerator[_T, None]:
    # blocking __exit__ from running waiting on a free thread
    # can create race conditions/deadlocks if the context manager itself
    # has it's own internal pool (e.g. a database connection pool)
    # to avoid this we let __exit__ run without a capacity limit
    # since we're creating a new limiter for each call, any non-zero limit
    # works (1 is arbitrary)
    exit_limiter = CapacityLimiter(1)
    try:
        yield await run_in_threadpool(cm.__enter__)
    except Exception as e:
        traceback_ = e.__traceback__
        ok = bool(
            await anyio.to_thread.run_sync(
                cm.__exit__, type(e), e, None, limiter=exit_limiter
            )
        )
        if not ok:
            if e.__traceback__ is None:
                e.__traceback__ = traceback_
            raise e
    else:
        await anyio.to_thread.run_sync(
            cm.__exit__, None, None, None, limiter=exit_limiter
        )

# Run this before starting anything FastAPI related
from fastapi.dependencies import utils
utils.contextmanager_in_threadpool = contextmanager_in_threadpool
```

**Review by odiseo0 — APPROVED** [review]

**Comment by sombek** [comment]
@tiangolo Any idea when can we merge this?

**Review by BilalAlpaslan — APPROVED** [review]

**Comment by etsvetanov** [comment]
Does this PR need additional work or help?

**Comment by Kludex** [comment]
Copied code from https://github.com/fastapi/fastapi/discussions/8428#discussioncomment-8113007, thanks @guillaumegenthial

File: fastapi/concurrency.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from contextlib import asynccontextmanager as asynccontextmanager
 from typing import AsyncGenerator, ContextManager, TypeVar
 
-import anyio
+import anyio.to_thread
```
Comment (Kludex) [inline fastapi/concurrency.py:4]: nitpick.

File: tests/test_exception_handlers.py [hunk 2]
```diff
@@ -65,3 +77,12 @@ def test_override_server_error_exception_response():
...
     assert response.json() == {"exception": "server-error"}
+
+
+def test_traceback_for_dependency_with_yield():
+    client = TestClient(app, raise_server_exceptions=True)
+    with pytest.raises(ValueError) as exc_info:
+        client.get("/dependency-with-yield")
+    last_frame = exc_info.traceback[-1]
+    assert str(last_frame.path) == __file__
+    assert last_frame.lineno == raise_value_error.__code__.co_firstlineno
```
Comment (Kludex) [inline tests/test_exception_handlers.py:88]: If you revert the changes, you can verify that the dependency will not be the last frame (or in any frame, because it will not be present).

**Comment by tiangolo** [comment]
Great job @sombek, thank you! :rocket: 

And thanks @Kludex for the help, tweaks, and bringing my attention to this on the private channels. :bow: 

This will be available in version 0.115.6 released in the next hours. :rocket:

File: tests/test_exception_handlers.py (hunk 2 above)
Comment (tiangolo) [reply tests/test_exception_handlers.py:88]: Awesome, great, I did this to verify. :nerd_face:
